satellite/repair/repairer/repairer.go

```diff
@@ -38,65 +40,106 @@ type Config struct {
 //
 // architecture: Worker
 type Service struct {
-	log      *zap.Logger
-	queue    queue.RepairQueue
-	config   *Config
-	Limiter  *sync2.Limiter
-	Loop     sync2.Cycle
-	repairer *SegmentRepairer
+	log        *zap.Logger
+	queue      queue.RepairQueue
+	config     *Config
+	JobLimiter *semaphore.Weighted
+	Loop       sync2.Cycle
+	repairer   *SegmentRepairer
 }
 
 // NewService creates repairing service
 func NewService(log *zap.Logger, queue queue.RepairQueue, config *Config, repairer *SegmentRepairer) *Service {
 	return &Service{
-		log:      log,
-		queue:    queue,
-		config:   config,
-		Limiter:  sync2.NewLimiter(config.MaxRepair),
-		Loop:     *sync2.NewCycle(config.Interval),
-		repairer: repairer,
+		log:        log,
+		queue:      queue,
+		config:     config,
+		JobLimiter: semaphore.NewWeighted(int64(config.MaxRepair)),
+		Loop:       *sync2.NewCycle(config.Interval),
+		repairer:   repairer,
 	}
 }
 
 // Close closes resources
 func (service *Service) Close() error { return nil }
 
+// WaitForPendingRepairs waits for all ongoing repairs to complete.
+//
+// NB: this assumes that service.config.MaxRepair will never be changed once this Service instance
+// is initialized. If that is not a valid assumption, we should keep a copy of its initial value to
+// use here instead.
+func (service *Service) WaitForPendingRepairs() {
+	// No error return is possible here; context.Background() can't be canceled
+	_ = service.JobLimiter.Acquire(context.Background(), int64(service.config.MaxRepair))
```

Comment:
why not take `ctx` as an arg to `WaitForPendingRepairs` and pass it into `Acquire` so that if we shut down the satellite the repairer does not deadlock?

Reply:
I was trying to make it a close match for the way `service.Limiter.Wait()` worked before. But yeah, that is probably a change worth making.

Reply:
Actually, looking at it, it's more complicated than it looks. If we added that then we'd need to handle error returns from Acquire(), which would mean WaitForPendingRepairs() would need to return an error, which would mean that error needs to be handled in all the places it's called (around 10). I'm not so sure it's worth it. If we are shutting down the satellite and we have canceled the context, all the `Service.process()` jobs holding slots will exit and release, after which this acquire call will succeed and return. If there are many calls to this function waiting, all of them will succeed in turn and return as well. So there shouldn't be possibility of deadlock still.